### PR TITLE
PACA-520: recognize rda api idle disconnects

### PR DESCRIPTION
**Pull request by brianburton**
**JIRA Ticket:**
[PACA-520](https://jira.cms.gov/browse/PACA-520)

**User Story or Bug Summary:**

After 10 mins of inactivity, the MPSM Kong gateway server disconnects the stream API connection, which results in an error on our side since we didn't expect the disconnect.  Our strategy going forward is that we should accept this and just try to reconnect in X mins.

---

### What Does This PR Do?

Recognizes these dropped connections and either:

1. if they happen after a configurable amount of idle time just logs and info message about the disconnect and shuts down normally, or
2. if they happen with less idle time passes through the exception normally.

Adds a new environment variable that can be used to override the default value of 4 minutes.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [x] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    4. There isn't too much of a burden on reviewers.
    5. Any problems it causes have a small "blast radius".
    6. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

**Review by ericbuckley — DISMISSED** [review]

**Review by KyRobbins — DISMISSED** [review]

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/GrpcRdaSource.java [hunk 1]
```diff
@@ -101,17 +109,21 @@ public GrpcRdaSource(
    */
   @VisibleForTesting
   GrpcRdaSource(
+      Clock clock,
```
Comment (lsmitchell) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/GrpcRdaSource.java:112]: Please update doc for the new fields here.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/GrpcRdaSource.java [hunk 2]
```diff
@@ -302,12 +350,17 @@ private Config(
...
+        @Nullable Duration minIdleTimeBeforeConnectionDrop,
+        @Nullable String authenticationToken) {
       this.serverType = Preconditions.checkNotNull(serverType, "serverType is required");
       this.host = host;
       this.port = port;
       this.inProcessServerName = inProcessServerName;
       this.maxIdle = maxIdle;
+      this.minIdleTimeBeforeConnectionDrop =
+          minIdleTimeBeforeConnectionDrop == null
+              ? Duration.ofMinutes(4)
```
Comment (lsmitchell) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/GrpcRdaSource.java:362]: Can this use the default specified in AppConfiguration so we have it defined only in one place?

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/GrpcRdaSource.java [hunk 3]
```diff
@@ -430,6 +483,10 @@ private Long parseJWTExpirationDate(String token) {
 
       return null;
     }
+
+    private long getMinIdleMillisBeforeConnectionDrop() {
```
Comment (lsmitchell) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/GrpcRdaSource.java:487]: Is this being used? It's private but I don't see it being referenced. (Also missing docs if it is being used)

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/GrpcResponseStream.java [hunk 4]
```diff
@@ -98,4 +127,19 @@ public synchronized StatusRuntimeException getCause() {
...
+  /**
+   * Wrapper for a {@code StatusRuntimeException} that was thrown in response to a stream reset
+   * caused by the server dropping its HTTP connection unexpectedly.
+   */
+  public static class DroppedConnectionException extends Exception {
+    public DroppedConnectionException(StatusRuntimeException cause) {
+      super(cause);
+    }
+
+    @Override
```
Comment (lsmitchell) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/GrpcResponseStream.java:140]: Should have an /* {@InheritDoc} */

**Review by lsmitchell — COMMENTED** [review]
Thanks for documenting and updating some of the tests in the touched classes!

Just a couple things, looks good otherwise.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/GrpcRdaSource.java (hunk 2 above)
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/GrpcRdaSource.java:362]: Cross module circular dependencies like that are undesirable.  Also, the default defined at the application level is used when constructing instances of this class so this class level default is never used in the application.  I'll just change it to Long.MAX_VALUE so the default is just to not use this new feature unless a value is supplied.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/GrpcRdaSource.java (hunk 3 above)
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/GrpcRdaSource.java:487]: Yes it's being used on line 97.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/GrpcResponseStream.java (hunk 4 above)
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/GrpcResponseStream.java:140]: Done.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/GrpcRdaSource.java (hunk 3 above)
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/GrpcRdaSource.java:494]: Comment added.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/GrpcRdaSource.java (hunk 2 above)
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/GrpcRdaSource.java]: Default changed.

**Review by lsmitchell — APPROVED** [review]

**Review by cbrunefearless — APPROVED** [review]
lgtm

**Review by KyRobbins — APPROVED** [review]

**Pull request merged**